UnchainedLauncher.Core/Services/Mods/Registry/ModRegistryFactory.cs

```diff
@@ -0,0 +1,46 @@
+using DiscriminatedUnions;
+using UnchainedLauncher.Core.Services.Mods.Registry.Downloader;
+using UnchainedLauncher.Core.Utilities;
+
+namespace UnchainedLauncher.Core.Services.Mods.Registry {
+    public class ModRegistryFactory {
+        public static readonly JsonFactory<ModRegistryMetadata, IModRegistry> Instance =
+            new JsonFactory<ModRegistryMetadata, IModRegistry>(ToClassType, ToJsonType);
+        
+        public static IModRegistry ToClassType(ModRegistryMetadata metadata) =>
+            metadata switch {
+                LocalModRegistryMetadata m => new LocalModRegistry(m.PakReleaseDir, ModRegistryDownloaderFactory.Instance.ToClassType(m.DownloaderMetadata)),
+                GithubModRegistryMetadata m => new GithubModRegistry(m.Org, m.RepoName, ModRegistryDownloaderFactory.Instance.ToClassType(m.DownloaderMetadata)),
+                AggregateModRegistryMetadata m => new AggregateModRegistry(m.Registries.Select(ToClassType).ToArray()),
+                
+                _ => throw new ArgumentOutOfRangeException(nameof(metadata), metadata,
+                    $"Attempt to initialize unknown IModRegistry implementation: {metadata}")
+            };
+        
+        public static ModRegistryMetadata ToJsonType(IModRegistry registry) =>
+            registry switch {
+                LocalModRegistry d => new LocalModRegistryMetadata(d.RegistryPath, ModRegistryDownloaderFactory.Instance.ToJsonType(d.ModRegistryDownloader)),
+                GithubModRegistry d => new GithubModRegistryMetadata(d.Organization, d.RepoName, ModRegistryDownloaderFactory.Instance.ToJsonType(d.ModRegistryDownloader)),
+                AggregateModRegistry d => new AggregateModRegistryMetadata(d.ModRegistries.Select(ToJsonType).ToArray()),
+                
+                _ => throw new ArgumentOutOfRangeException(nameof(registry), registry,
+                    $"Attempt to extract metadata from unknown IModRegistry implementation: {registry}")
```

Comment:
This method and the other method above both throw rather than using safer error handling, because these methods are things that should be done at the beginning and end of the application lifecycle, where crashes are more "acceptable"